### PR TITLE
Added TokenOf

### PR DESCRIPTION
**Pull request by Superstar64**
Added TokenOf template to the lexer and modifyed files that used TokenType.

### CODE REVIEW

**Comment by schancel** [comment]
Interesting idea.  Not sure how I feel about it.  What's your reasoning behind this change @Superstar64 ?

**Comment by Superstar64** [comment]
It looks more readable(In my opinion).

**Comment by deadalnix** [comment]
I'm not sure about this change. It create a lot of changes but do not bring much at the end.

Arguably, TokenOf!"X" may be more readable than TokenType.X but this diff do not even allow to translate all existing code, and leave us with 2 ways of doing the same thing.

@schancel do you think this is more readable ? I'm not really against this change but we got to go one way or another. Half way in between is worse than the 2 options, IMO.

**Comment by schancel** [comment]
I like that it creates a 1-1 map between what you would type as far as D-code.  However, I dislike that it breaks my "intellisense" in my IDE. :)

I would be in support of the change under two conditions. 

1) All code was stylized in this manner.
2) TokenOf!"\0" does not get used.  I prefer special symbols for begin and end.  e.g. `TokenType.Begin` and `TokenType.End` should be defined.

**Comment by Superstar64** [comment]
Ok, I reverted TokenOf!"\0".
What exact did I miss, I made and used a small D program to recursive replace everything in src/.

**Comment by deadalnix** [comment]
Ok I have nothign against this change, and if @schancel think it is more readable as well, we can go in that direction. But I have something against having 2 way of doing the same thing.

Either move everything to use TokenOf!"X" or nothing (in which case you could rename TokenOf to TokenType as the first one is gone).

**Comment by Superstar64** [comment]
You can't replace TokenType with TokenOf because TokenType needs to be able to be used as a Type.

I Replaced all uses of TokenType with TokenOf but I'm not sure if this is how you wanted it.

**Comment by Superstar64** [comment]
I also found a bug,At 
https://github.com/deadalnix/libd/blob/0cc4155d9c7da050520edc48ff9e0550962b26e2/src/d/parser/util.d#L18
TokenType.Greater doesn't exist.

**Comment by Superstar64** [comment]
I realized that I missed all the with(TokenType) statements and fixed them.

**Comment by Superstar64** [comment]
Sorry,Accidentally click close and comment instead of comment.

**Comment by UplinkCoder** [comment]
I see the issue with code-completion.
If every `TokenType` would have a ddoc comment one would see the string-representation directly.
And template-instantiations are not for free.

File: src/d/lexer.d [hunk 1]
```diff
@@ -870,183 +760,316 @@ mixin template CharPumper(bool decode = true) {
...
+				return *result;
+			}
+			assert(0,"Can't Find Token");
+		}
+	}
+	enum TokenOf=genVal;
+}
+
+
 auto getOperatorsMap() {
```
Comment (deadalnix) [inline src/d/lexer.d]: If this is using tokenof

File: src/d/lexer.d [hunk 2]
```diff
@@ -870,183 +760,316 @@ mixin template CharPumper(bool decode = true) {
...
+}
+
 public:
+template TokenOf(string token){
+	auto genVal(){
+		static if(__traits(hasMember, TokenType_, token)){
+			return __traits(getMember, TokenType_, token);
+		}else{
+			TokenType* result;
+			result=token in getOperatorsMap;
```
Comment (deadalnix) [inline src/d/lexer.d]: And token is using this, that do not seems like a sound design.

**Comment by deadalnix** [comment]
Ok overall I like the direction where this is going. I understand that template are not free and all, but the whole thing should be instantiated once and never dealt with again, so that shouldn't be too big of an issue.

Now more on the matter. I don't like he name TokenOf that much as it is telling the reader that he should get a token. But he gets a TokenType, so this should really be TokenTypeOf (unless someone has a better name ?).

The change from TokenType to TokenType_ seems unnecessary. Making it private is enough.

Also, the added code does not follow the style. Please put spacing around = or before/after curly braces. Do not use optional ().

File: src/d/lexer.d [hunk 3]
```diff
@@ -870,183 +760,316 @@ mixin template CharPumper(bool decode = true) {
...
+	version(KIRBY) {
+		KirbyRape,		// (>^(>O_O)>
+	}
+	*/
+}
+
 public:
+template TokenOf(string token){
+	auto genVal(){
+		static if(__traits(hasMember, TokenType_, token)){
```
Comment (deadalnix) [inline src/d/lexer.d]: This will cause all TokenOf for keywords and/or operators to be duplicated.

**Comment by UplinkCoder** [comment]
TokenTypeOf is fine.

**Comment by Superstar64** [comment]
TokenType had to be renamed to to TokenType_ because the lexer needs a public TokenType symbol (here https://github.com/Superstar64/libd/blob/b8fc4605fb57c21fc5ef2ab8d15bb81f3aeb6fb9/src/d/lexer.d#L14). Having both would cause a symbol conflict

**Comment by deadalnix** [comment]
That effectively expose TokenType_, so the stunt seems useless.

**Comment by Superstar64** [comment]
Ok, i reverted it. Do you know of anyway to use TokenType as a Type with out allow access to it's elements?

**Comment by deadalnix** [comment]
It seems that it is not supported for enums :(

Then I guess it is ok to expose it if we never use it directly (for consistency).

**Comment by Superstar64** [comment]
Should typedef be added to getKeywordsMap(here: https://github.com/deadalnix/libd/blob/59fe63a5ad84aaf53faa73bed710ac94abd84529/src/d/lexer.d#L948)? There is one instance of TokenType.Typedef here:https://github.com/deadalnix/libd/blob/59fe63a5ad84aaf53faa73bed710ac94abd84529/src/d/parser/expression.d#L883

**Comment by deadalnix** [comment]
Yes it should be added (however, we shouldn't spend any time on this).

**Comment by deadalnix** [comment]
Thinking about this, the gain appears less and less clearly to me.

Sure, TokenTypeOf!">=" is clearer, but TokenTypeOf!"static" is certainly more cumbersome then TokenType.Static, especially with with statements.

**Comment by Superstar64** [comment]
How do you suggest improving it then?

**Comment by deadalnix** [comment]
OK the current list of token is annoying because naming is inconsistent. Some token are named after what they are (OpenParen) and some are named after what they do in D (Assign). Maybe making the naming more consistent could make things easier ?

**Comment by Superstar64** [comment]
Is there anything stopping this pull request from getting pulled/rejected?

**Comment by deadalnix** [comment]
I don't think this is a good idea anymore, for above mentioned reasons:
 - It is better for operator, but worse for keywords
 - The confusion caused on operator can be mitigated using consistent naming.

I wanted to have feedback on this before closing as I may have overlooked something.

**Comment by UplinkCoder** [comment]
I'd say cherry-pick https://github.com/Superstar64/libd/commit/b94fab970a1e16e0dc2f7c16a5b22ac85ee51a4e

**Comment by schancel** [comment]
Yeah, I'm not a huge fan.  I do appreciate the work though @Superstar64.   

Lets get `b94fab9` in and fix the typo.

**Comment by Superstar64** [comment]
Should I close this pull request and open a new one?

**Comment by deadalnix** [comment]
As you see fit, either open a new one and update this one. The commit mentioned by @UplinkCoder seems to be like we want (modulo some typos).

**Comment by Superstar64** [comment]
Alright, New pull request it is.

**Pull request closed without merging**